build-tools/src/main/java/org/elasticsearch/gradle/OS.java

```diff
@@ -58,8 +58,7 @@ public Conditional<T> onUnix(Supplier<T> supplier) {
         }
 
         public T supply() {
-            HashSet<OS> missingOS = new HashSet<>(Arrays.asList(OS.values()));
-            missingOS.removeAll(conditions.keySet());
+            Set<OS> missingOS = EnumSet.complementOf(EnumSet.copyOf(conditions.keySet()));
```

Comment:
Annoyingly, `EnumMap.keySet()` is not itself an `EnumSet` (although it should be). So I'm not sure that this is more efficient than just creating an `EnumSet.allOf` and then doing a `removeAll` - only one object is created then

Reply:
The first `keySet()` call also creates a new object. And then `removeAll()` will also create an `Iterator` for that Set.
Whereas `enumSet.complement()` just inverts the bit mask.

Reply:
Yes, but `complementOf` also creates a copy of the enumset - so you're copying it twice. The iterator inside `removeAll` is a smaller object and likely to be optimised onto the stack, which `copyOf`/`complementOf` is less likely to be as it is returned from a method call.

It's also clearer to create the set then do a `removeAll`.